src/test/java/ir/sahab/nexus/plugin/tag/internal/dto/IntegrationTest.java

```diff
@@ -0,0 +1,222 @@
+        output.addFormData("maven2.version", component.getVersion(), MediaType.TEXT_PLAIN_TYPE);
+        output.addFormData("maven2.asset1.extension", "jar", MediaType.TEXT_PLAIN_TYPE);
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(randomAlphanumeric(100).getBytes());
+        output.addFormData("maven2.asset1", inputStream, MediaType.APPLICATION_OCTET_STREAM_TYPE,
+                "test-maven-artifact.jar");
+
+        Response response = target.path("components").queryParam("repository", component.getRepository())
+                .request()
+                .header(HttpHeaders.AUTHORIZATION, authorizationHeader())
+                .post(Entity.entity(output, MediaType.MULTIPART_FORM_DATA_TYPE));
+        assertEquals(Family.SUCCESSFUL, response.getStatusInfo().getFamily());
+    }
+
+    /**
+     * @return authorization header of basic authentication
+     */
+    private static String authorizationHeader() {
+        String auth = USERNAME + ":" + PASSWORD;
+        byte[] encodedAuth = Base64.encodeBase64(auth.getBytes(StandardCharsets.ISO_8859_1));
+        return "Basic " + new String(encodedAuth);
+    }
+
+    @Test
+    public void testCrud() {
+        Map<String, String> attributes = new HashMap<>();
+        attributes.put(CHANGE_ID, randomAlphanumeric(20));
+        attributes.put(STATUS, "failed");
+        attributes.put("Commit-Id", randomAlphanumeric(20));
+
+        TagDefinition tag = new TagDefinition(randomAlphanumeric(5), attributes, Arrays.asList(component1));
+
+        // Add Tag
+        Response response = addTag(tag);
+        assertEquals(Family.SUCCESSFUL, response.getStatusInfo().getFamily());
+        Tag postResponseTag = response.readEntity(Tag.class);
+        assertFalse(new Date().before(postResponseTag.getFirstCreated()));
+        assertEquals(postResponseTag.getFirstCreated(), postResponseTag.getLastUpdated());
+        assertDefinitionEquals(tag, postResponseTag);
+
+        // Test Get by name
+        Tag retrieved = target.path("tags/" + tag.getName()).request().get(Tag.class);
+        assertDefinitionEquals(tag, retrieved);
+
+        // Test search by attribute
+        List<Tag> result = target.path("tags")
+                .queryParam("attributes", CHANGE_ID + "=" + tag.getAttributes().get(CHANGE_ID))
+                .queryParam("attributes", STATUS + "=" + tag.getAttributes().get(STATUS))
+                .request()
+                .get(new GenericType<List<Tag>>() {});
+        assertEquals(1, result.size());
+        assertDefinitionEquals(tag, result.get(0));
+
+        // Update tag
+        tag.getAttributes().put(STATUS, "successful");
+        tag.setComponents(Arrays.asList(component1, component2));
+        response = updateTag(tag);
+        Tag putResponseTag = response.readEntity(Tag.class);
+        assertEquals(postResponseTag.getFirstCreated(), putResponseTag.getFirstCreated());
+        assertFalse(new Date().before(putResponseTag.getLastUpdated()));
```

Comment:
assert lastUpdated is after firstCreated

Reply:
Done